modules/gdscript/gdscript_compiler.cpp

```diff
@@ -1059,22 +1059,22 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 					bool known_type = assigned.type.has_type;
 					bool is_shared = Variant::is_type_shared(assigned.type.builtin_type);
 
-					if (!known_type) {
-						// Jump shared values since they are already updated in-place.
-						gen->write_jump_if_shared(assigned);
-					}
-					if (known_type && !is_shared) {
+					if (!known_type || !is_shared) {
+						if (!known_type) {
+							// Jump shared values since they are already updated in-place.
+							gen->write_jump_if_shared(assigned);
+						}
 						if (!info.is_named) {
 							gen->write_set(info.base, info.key, assigned);
-							if (info.key.mode == GDScriptCodeGenerator::Address::TEMPORARY) {
-								gen->pop_temporary();
-							}
 						} else {
 							gen->write_set_named(info.base, info.name, assigned);
 						}
+						if (!known_type) {
+							gen->write_end_jump_if_shared();
+						}
 					}
-					if (!known_type) {
-						gen->write_end_jump_if_shared();
+					if (key.mode == GDScriptCodeGenerator::Address::TEMPORARY) {
```

Comment:
I guess this does need to be moved out of the `if`, but it needs to be checked if it's not named (if it is, then it doesn't have `key`) and it's the `info.key` since standalone`key` is something else.

```suggestion
					if (!info.is_named && info.key.mode == GDScriptCodeGenerator::Address::TEMPORARY) {
```